### PR TITLE
Dsprites

### PR DESCRIPTION
**Pull request by ysharma1126**

### CODE REVIEW

**Comment by dpaiton** [comment]
Thanks for doing this! A few requests:

- [ ] Notebooks: Not sure if you are doing this, but please make sure you "Clear Output" and save all notebooks before committing them.

- [ ] train_model.py: Please remove thisfrom the PR. I'd like to keep this file as a general example that is not specific to any one model type.

- [ ] dataset_utils.py:
    - [ ] Please make `root` an extension of `params.data_dir`, as is done with MNIST. This ensures that the data is located where the user wants it. For example: `root = os.path.join(*[params.data_dir, 'dsprites-dataset', 'dsprites_ndarray_co1sh3sc6or40x32y32_64x64.npz`]).
    - [ ] Change line 44 to say 'Now downloading the dsprites-dataset...'
    - [ ] The test & val split code is admittedly clunky, and now I think we need to clean it up. I have a suggestion, but I'm open to discussion if you don't like it. Any dataset should be able to be split into a validation & test set, even if it is procedurally generated or does not normally have labels. Whether or not we do this should be specified in the params. Right now the code (lines 63:67) allows for `params.num_[val/test]_images` to not exist, and it sets it if it doesn't exist. Let's throw that out, make the parameter mandatory (integer or None) & default it in base_params.py with `None`. For each dataset, we check `params.num_val_images` and `params.num_test_images`. If they are None or 0, then we set the loaders to None. Any other integer would indicate partitioning a random set of the dataset into the corresponding loaders. For dsprites, this would be done before setting up the `train_loader`, i.e. do the split on `train_data`. MNIST right now only has train/test. To fix this I think we need to use our own data loader from the MNIST npz files, much in the same way that you setup dsprites. I am happy to make MNIST my job for after we merge this PR. 

- [ ] download_dsprites.sh: Like above, please make the download location an extension of params.data_dir. You'll need to use function args for this.

- [ ] run_utils.py: I like the idea of adding `return_pics`, however I'd like it to be more general. It basically is saying return whatever the output to the model's forward pass is, right? So instead of pics can we say something like `return_model_outputs`?
